apollo-router/src/metrics/aggregation.rs

```diff
@@ -0,0 +1,382 @@
+// This is within the spec: https://opentelemetry.io/docs/specs/otel/metrics/api/#get-a-meter
+// `Meters are identified by name, version, and schema_url fields. When more than one Meter of the same name, version, and schema_url is created, it is unspecified whether or under which conditions the same or different Meter instances are returned. It is a user error to create Meters with different attributes but the same identity.`
+
+#[derive(Hash, Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Debug)]
+pub(crate) enum MeterProviderType {
+    Prometheus,
+    Apollo,
+    General,
+}
+
+#[derive(Clone, Default)]
+pub(crate) struct AggregateMeterProvider {
+    inner: Arc<Mutex<Inner>>,
+}
+
+#[derive(Default)]
+struct Inner {
+    providers: HashMap<MeterProviderType, (FilterMeterProvider, HashMap<MeterId, Meter>)>,
+}
+
+#[derive(Eq, PartialEq, Hash)]
+struct MeterId {
+    name: Cow<'static, str>,
+    version: Option<Cow<'static, str>>,
+    schema_url: Option<Cow<'static, str>>,
+    // Note that attributes are not part of the meter ID.
+}
+
+impl AggregateMeterProvider {
+    /// The behaviour of this function is that if None is passed in, the meter will be left as is.
+    /// To disable meter_providers use a noop meter provider.
+    /// The old meter_provider if any is returned, and it is up to the caller to clean up.
+    pub(crate) fn set(
+        &self,
+        meter_provider_type: MeterProviderType,
+        meter_provider: Option<FilterMeterProvider>,
+    ) -> Option<FilterMeterProvider> {
+        let mut inner = self.inner.lock().expect("lock poisoned");
+        if let Some(meter_provider) = meter_provider {
+            inner
+                .providers
+                .insert(
+                    meter_provider_type,
+                    (meter_provider.clone(), HashMap::new()),
+                )
+                .map(|(old_provider, _)| old_provider)
+        } else {
+            None
+        }
+    }
+
+    /// Shutdown MUST be called from a blocking thread.
+    pub(crate) fn shutdown(&self) {
+        let inner = self.inner.lock().expect("lock poisoned");
+        for (meter_provider_type, (meter_provider, _)) in &inner.providers {
+            if let Err(e) = meter_provider.shutdown() {
+                ::tracing::error!(error = %e, meter_provider_type = ?meter_provider_type, "failed to shutdown meter provider")
+            }
+        }
+    }
+}
+
+impl MeterProvider for AggregateMeterProvider {
+    fn versioned_meter(
+        &self,
+        name: impl Into<Cow<'static, str>>,
+        version: Option<impl Into<Cow<'static, str>>>,
+        schema_url: Option<impl Into<Cow<'static, str>>>,
+        attributes: Option<Vec<KeyValue>>,
+    ) -> Meter {
+        let name = name.into();
+        let version = version.map(|v| v.into());
+        let schema_url = schema_url.map(|v| v.into());
+
+        let mut meters = Vec::new();
+        let mut inner = self.inner.lock().expect("lock poisoned");
```

Comment:
this is a sync mutex, and we have a lot of metrics, so it is highly likely that executors would have contention over it. The opentelemetry crate uses a sync mutex too, but shifts the work elsewhere https://github.com/open-telemetry/opentelemetry-rust/blob/3ee22ced04c17b54cabe6ae52ef90a55b37fa5cf/opentelemetry-sdk/src/metrics/meter_provider.rs#L124-L132

Reply:
Going to resolve this as callsite caching has been implemented and this code is not on the hot path.